arduino/libraries/librarieslist.go

```diff
@@ -41,14 +41,18 @@ func (list *List) Add(libs ...*Library) {
 	}
 }
 
-// Remove removes the library from the list
-func (list *List) Remove(library *Library) {
+// Remove removes the given library from the list
+func (list *List) Remove(libraryToRemove *Library) {
+	j := 0
 	for i, lib := range *list {
-		if lib == library {
-			*list = append((*list)[:i], (*list)[i+1:]...)
-			return
+		if lib != libraryToRemove {
+			if i != j {
+				(*list)[j] = (*list)[i]
+			}
+			j++
 		}
 	}
+	*list = (*list)[:j]
```

Comment:
```
goos: linux
goarch: amd64
pkg: prova
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
                                                 │   old.txt   │               new.txt                │
                                                 │   sec/op    │    sec/op     vs base                │
RemoveOldBestCaseScenario/input_size_100-20        171.0n ± 2%    232.2n ± 2%  +35.76% (p=0.000 n=10)
RemoveOldBestCaseScenario/input_size_1000-20       1.435µ ± 1%    2.261µ ± 3%  +57.53% (p=0.000 n=10)
RemoveOldBestCaseScenario/input_size_5000-20       6.979µ ± 1%   10.953µ ± 3%  +56.94% (p=0.000 n=10)
RemoveOldBestCaseScenario/input_size_10000-20      13.61µ ± 1%    21.31µ ± 3%  +56.62% (p=0.000 n=10)
RemoveOldBestCaseScenario/input_size_100000-20     176.9µ ± 2%    249.3µ ± 4%  +40.99% (p=0.000 n=10)
RemoveOldMedianCaseScenario/input_size_100-20      177.3n ± 1%    221.9n ± 3%  +25.19% (p=0.000 n=10)
RemoveOldMedianCaseScenario/input_size_1000-20     1.615µ ± 3%    2.042µ ± 3%  +26.45% (p=0.000 n=10)
RemoveOldMedianCaseScenario/input_size_5000-20     8.146µ ± 3%    9.699µ ± 4%  +19.06% (p=0.000 n=10)
RemoveOldMedianCaseScenario/input_size_10000-20    15.18µ ± 2%    19.18µ ± 3%  +26.38% (p=0.000 n=10)
RemoveOldMedianCaseScenario/input_size_100000-20   163.6µ ± 2%    205.4µ ± 2%  +25.57% (p=0.000 n=10)
RemoveOldWorstCaseScenario/input_size_100-20       183.2n ± 1%    187.7n ± 3%   +2.46% (p=0.000 n=10)
RemoveOldWorstCaseScenario/input_size_1000-20      1.650µ ± 2%    1.813µ ± 2%   +9.91% (p=0.000 n=10)
RemoveOldWorstCaseScenario/input_size_5000-20      7.768µ ± 1%    8.532µ ± 2%   +9.84% (p=0.000 n=10)
RemoveOldWorstCaseScenario/input_size_10000-20     15.10µ ± 1%    16.75µ ± 3%  +10.89% (p=0.000 n=10)
RemoveOldWorstCaseScenario/input_size_100000-20    158.1µ ± 3%    174.8µ ± 2%  +10.59% (p=0.000 n=10)
geomean                                            5.517µ         6.974µ       +26.41%

```
I was curious to test the new implementation and I ran some benchmarks. I created some synthetic data and run multiple sizes and searched for best/median/worst-case scenarios.
It seems to perform slower than the old implementation.

Besides raw performances is there any reason why we opted for this slightly different implementation?

---
This is how I did the benchmark:
```
$ go test -bench=Old -count 10 | tee old.txt

Then I swapped the implementation to use the new one and run:
$ go test -bench=Old -count 10 | tee new.txt

To check the diff I run:
$ go install golang.org/x/perf/cmd/benchstat@latest
$ benchstat old.txt new.txt
```
```go
package main

import (
    "fmt"
	"testing"
)
type Library struct {
	Name          string
	Author        string
	Maintainer    string
	Sentence      string
	Paragraph     string
	Website       string
	Category      string
	Architectures []string

	Types []string `json:"types,omitempty"`

	DirName                string
	DotALinkage            bool
	Precompiled            bool
	PrecompiledWithSources bool
	LDflags                string
	IsLegacy               bool
	InDevelopment          bool
	License                string
	declaredHeaders        []string
	sourceHeaders          []string
	CompatibleWith         map[string]bool
}

type List []*Library

func main() {
}

func generateList(l int) List {
	a := make(List, l)
	for i := 1; i <= l; i++ {
		a[i-1] = &Library{
			Name:                   fmt.Sprintf("name%d", i),
			Author:                 fmt.Sprintf("author%d", i),
			Maintainer:             fmt.Sprintf("maintainer%d", i),
			Sentence:               fmt.Sprintf("sentence%d", i),
			Paragraph:              fmt.Sprintf("paragraph%d", i),
			Website:                fmt.Sprintf("website%d", i),
			Category:               fmt.Sprintf("category%d", i),
			Architectures:          []string{"avr"},
			Types:                  []string{"type"},
			DirName:                "dirname",
			DotALinkage:            false,
			Precompiled:            false,
			PrecompiledWithSources: false,
			LDflags:                "--flag",
			IsLegacy:               false,
			InDevelopment:          false,
			License:                "license",
			declaredHeaders:        []string{"header1"},
			sourceHeaders:          []string{"header1"},
			CompatibleWith:         map[string]bool{"a": true},
		}
	}
	return a
}

func (list *List) Remove(library *Library) {
	for i, lib := range *list {
		if lib == library {
			*list = append((*list)[:i], (*list)[i+1:]...)
			return
		}
	}
}

func (list *List) Remove2(libraryToRemove *Library) {
	j := 0
	for i, lib := range *list {
		if lib != libraryToRemove {
			if i != j {
				(*list)[j] = (*list)[i]
			}
			j++
		}
	}
	*list = (*list)[:j]
}

var table = []struct {
	input int
	list  List
}{
	{input: 100, list: generateList(100)},
	{input: 1000, list: generateList(1000)},
	{input: 5000, list: generateList(5000)},
	{input: 10000, list: generateList(10000)},
	{input: 100000, list: generateList(100000)},
}

func BenchmarkRemoveOldBestCaseScenario(b *testing.B) {
	for _, v := range table {
		b.Run(fmt.Sprintf("input_size_%d", v.input), func(b *testing.B) {
			bestCaseTarget := 0
			for i := 0; i < b.N; i++ {
				a := make(List, v.input)
				copy(a, v.list)
				a.Remove(a[bestCaseTarget])
			}
		})
	}
}

func BenchmarkRemoveOldMedianCaseScenario(b *testing.B) {
	for _, v := range table {
		b.Run(fmt.Sprintf("input_size_%d", v.input), func(b *testing.B) {
			meanCaseTarget := v.input / 2
			for i := 0; i < b.N; i++ {
				a := make(List, v.input)
				copy(a, v.list)
				a.Remove(a[meanCaseTarget])
			}
		})
	}
}
func BenchmarkRemoveOldWorstCaseScenario(b *testing.B) {
	for _, v := range table {
		b.Run(fmt.Sprintf("input_size_%d", v.input), func(b *testing.B) {
			worstCaseTarget := v.input - 1
			for i := 0; i < b.N; i++ {
				a := make(List, v.input)
				copy(a, v.list)
				a.Remove(a[worstCaseTarget])
			}
		})
	}
}
func BenchmarkRemoveNewBestCaseScenario(b *testing.B) {
	for _, v := range table {
		b.Run(fmt.Sprintf("input_size_%d", v.input), func(b *testing.B) {
			bestCaseTarget := 0
			for i := 0; i < b.N; i++ {
				a := make(List, v.input)
				copy(a, v.list)
				a.Remove2(a[bestCaseTarget])
			}
		})
	}
}

func BenchmarkRemoveNewMedianCaseScenario(b *testing.B) {
	for _, v := range table {
		b.Run(fmt.Sprintf("input_size_%d", v.input), func(b *testing.B) {
			meanCaseTarget := v.input / 2
			for i := 0; i < b.N; i++ {
				a := make(List, v.input)
				copy(a, v.list)
				a.Remove2(a[meanCaseTarget])
			}
		})
	}
}
func BenchmarkRemoveNewWorstCaseScenario(b *testing.B) {
	for _, v := range table {
		b.Run(fmt.Sprintf("input_size_%d", v.input), func(b *testing.B) {
			worstCaseTarget := v.input - 1
			for i := 0; i < b.N; i++ {
				a := make(List, v.input)
				copy(a, v.list)
				a.Remove2(a[worstCaseTarget])
			}
		})
	}
}
```

Reply:
Thanks for making the benchmarks!
Actually, I don't remember where this change comes from, it was made in the past and I've just blindly rebased it.
But at this point I'm going to revert it to the old better implementation...